website/source/guides/writing-custom-terraform-providers.html.md

```diff
@@ -0,0 +1,553 @@
+How do we properly tell Terraform to only persist half the diff? This is known
+as a _partial state_ scenario, and implementing these properly is critical to a
+well-behaving provider.
+
+Here are the rules for state updating in Terraform. Note that this mentions
+callbacks we have not discussed, for the sake of completeness.
+
+- If the `Create` callback returns with or without an error without an ID set
+  using `SetId`, the resource is assumed to not be created, and no state is
+  saved.
+
+- If the `Create` callback returns with or without an error and an ID has been
+  set, the resource is assumed created and all state is saved with it. Repeating
+  because it is important: if there is an error, but the ID is set, the state is
+  fully saved.
+
+- If the `Update` callback returns with or without an error, the full state is
+  saved. If the ID becomes blank, the resource is destroyed (even within an
+  update, though this shouldn't happen except in error scenarios).
+
+- If the `Destroy` callback returns without an error, the resource is assumed to
+  be destroyed, and all state is removed.
+
+- If the `Destroy` callback returns with an error, the resource is assumed to
+  still exist, and all prior state is preserved.
+
+- If partial mode (covered next) is enabled when a create or update returns,
+  only the explicitly enabled configuration keys are persisted, resulting in a
+  partial state.
+
+_Partial mode_ is a mode that can be enabled by a callback that tells Terraform
+that it is possible for partial state to occur. When this mode is enabled, the
+provider must explicitly tell Terraform what is safe to persist and what is not.
+
+Here is an example of a partial mode with an update function:
+
+```go
+func resourceServerUpdate(d *schema.ResourceData, m interface{}) error {
+	// Enable partial state mode
+	d.Partial(true)
+
+	if d.HasChange("address") {
+		// Try updating the address
+		if err := updateAddress(d, meta); err != nil {
+			return err
+		}
+
+		d.SetPartial("address")
+	}
+
+	// If we were to return here, before disabling partial mode below,
+	// then only the "address" field would be saved.
+
+	// We succeeded, disable partial mode. This causes Terraform to save
+	// save all fields again.
+	d.Partial(false)
+
+	return nil
+}
+```
+
+Note - this code will not compile since there is no `updateAddress` function.
+You can implement a dummy version of this function to play around with partial
+state. For this example, partial state does not mean much in this documentation
+example. If `updateAddress` were to fail, then the address field would not be
+updated.
+
+## Implementing Destroy
+
+The `Destroy` callback is exactly what it sounds like - it is called to destroy
+the resource. This operation should never update any state on the resource. It
+is not necessary to call `d.SetId("")`, since any non-error return value assumes
+the resource was deleted successfully.
```

Comment:
might make a note that it is not necessary, but shown here to clearly demonstrate(?)